src/System.Net.Http/src/System/Net/Http/HPack/HPackDecoder.cs

```diff
@@ -0,0 +1,398 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace System.Net.Http.HPack
+{
+    internal class HPackDecoder
+    {
+        // Can't use Action<,> because of Span
+        public delegate void HeaderCallback(ReadOnlySpan<byte> headerName, ReadOnlySpan<byte> headerValue);
+
+        private enum State
+        {
+            Ready,
+            HeaderFieldIndex,
+            HeaderNameIndex,
+            HeaderNameLength,
+            HeaderNameLengthContinue,
+            HeaderName,
+            HeaderValueLength,
+            HeaderValueLengthContinue,
+            HeaderValue,
+            DynamicTableSizeUpdate
+        }
+
+        public const int DefaultHeaderTableSize = 4096;
+
+        // TODO: add new configurable limit
```

Comment:
Are you going to open issues for each of these TODOs?  It'd be good to include whatever issue numbers you end up with in these comments.

Reply:
That's my intention, yes. At worst they will go in a followup PR.